backend/src/main/java/ai/verta/modeldb/lineage/LineageDAORdbImpl.java

```diff
@@ -30,36 +49,241 @@
+      lineageEntries.addAll(addLineage.getOutputList());
+      existsCheckConsumer.test(session, lineageEntries);
       for (LineageEntry input : addLineage.getInputList()) {
-        for (LineageEntry output : addLineage.getOutputList()) {
-          addLineage(session, input, output);
-        }
+        addLineage(session, input, id, CONNECTION_TYPE_INPUT, blobHashInCommitFunction);
+      }
+      for (LineageEntry output : addLineage.getOutputList()) {
+        addLineage(session, output, id, CONNECTION_TYPE_OUTPUT, blobHashInCommitFunction);
       }
       session.getTransaction().commit();
+      return AddLineage.Response.newBuilder().setId(lineageElementEntity.getId()).build();
     }
-    return AddLineage.Response.newBuilder().setStatus(true).build();
   }
 
   @Override
-  public DeleteLineage.Response deleteLineage(DeleteLineage deleteLineage) throws ModelDBException {
+  public DeleteLineage.Response deleteLineage(
+      DeleteLineage deleteLineage, BlobHashInCommitFunction blobHashInCommitFunction)
+      throws ModelDBException {
+    boolean status;
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
       session.beginTransaction();
       validate(deleteLineage.getInputList(), deleteLineage.getOutputList());
+      long id = deleteLineage.getId();
+      if (id == 0) {
+        throw new ModelDBException("Id not specified", Code.INVALID_ARGUMENT);
+      }
+      List connectionEntitiesById = getConnectionEntitiesById(session, id);
+      Map.Entry<Map<LineageElement, ConnectionEntity>, Map<LineageElement, ConnectionEntity>>
+          inputOutputs = getInputOutputs(session, connectionEntitiesById);
+      Map<LineageElement, ConnectionEntity> inputInDatabase = inputOutputs.getKey();
+      Map<LineageElement, ConnectionEntity> outputInDatabase = inputOutputs.getValue();
       for (LineageEntry input : deleteLineage.getInputList()) {
-        for (LineageEntry output : deleteLineage.getOutputList()) {
-          deleteLineage(session, input, output);
+        deleteLineage(session, input, inputInDatabase, blobHashInCommitFunction);
+      }
+      for (LineageEntry output : deleteLineage.getOutputList()) {
+        deleteLineage(session, output, outputInDatabase, blobHashInCommitFunction);
+      }
+      status = false;
+      boolean noDataProvided =
+          deleteLineage.getInputCount() == 0 && deleteLineage.getOutputCount() == 0;
+      if (noDataProvided || inputInDatabase.isEmpty() || outputInDatabase.isEmpty()) {
+        Set<ConnectionEntity> values = new HashSet<>(inputInDatabase.values());
+        values.addAll(outputInDatabase.values());
+        values.forEach(connectionEntity -> deleteConnectionEntity(session, connectionEntity));
+        LineageElementEntity lineageElementEntity = session.get(LineageElementEntity.class, id);
+        if (lineageElementEntity != null) {
+          session.remove(lineageElementEntity);
+          status = true;
         }
       }
```

Comment:
What is this doing? If it does what I'm thinking, it seems wrong.

Reply:
noDataProvided is a flag that there are no entries provided in the request. That means that we should delete all entries with the specified id. Is it not right?
Also when there are no input or output in the database left we should delete the remaining ones because they are not connected with anything.